sockeye/loss.py

```diff
@@ -110,33 +128,58 @@ def get_loss(self, logits: mx.sym.Symbol, labels: mx.sym.Symbol) -> List[mx.sym.
                                     name=C.SOFTMAX_NAME)]
 
 
-def _normalize(loss: mx.sym.Symbol, labels: mx.sym.Symbol):
+    def create_metric(self) -> "CrossEntropyMetric":
+        return CrossEntropyMetric(self.loss_config)
+
+
+class CrossEntropyMetric(EvalMetric):
     """
-    Normalize loss by the number of non-PAD tokens.
+    Version of the cross entropy metric that ignores padding tokens.
 
-    :param loss: A loss value for each label.
-    :param labels: A label for each loss entry (potentially containing PAD tokens).
-    :return: The normalized loss.
+    :param loss_config: The configuration used for the corresponding loss.
+    :param name: Name of this metric instance for display.
+    :param output_names: Name of predictions that should be used when updating with update_dict.
+    :param output_labels: Name of labels that should be used when updating with update_dict.
     """
-    return mx.sym.broadcast_div(loss, mx.sym.sum(labels != C.PAD_ID))
+    def __init__(self,
+                 loss_config: LossConfig,
+                 name: str = C.CROSS_ENTROPY,
+                 output_names: Optional[List[str]] = None,
+                 label_names: Optional[List[str]] = None) -> None:
+        super().__init__(name, output_names=output_names, label_names=label_names)
+        self.loss_config = loss_config
+
+    def update(self, labels, preds):
+        for label, pred in zip(labels, preds):
+            label = label.as_in_context(pred.context).reshape((label.size,))
+            prob = mx.nd.pick(pred, label.astype(dtype="int32"))
+            # Ignore padding
+            ignore = (label == C.PAD_ID).astype(dtype=prob.dtype)
+            prob = prob * (1 - ignore) + ignore
+            # Sum, normalizing if needed
+            loss = -mx.nd.log(prob + 1e-8)
+            if self.loss_config.normalize:
+                loss = loss / mx.nd.sum(1 - ignore)
+                self.num_inst += 1
+            else:
+                self.num_inst += label.size - mx.nd.sum(ignore).asscalar()
```

Comment:
This metric should reflect what the corresponding loss is doing, no? When we have `normalize=False` then the loss is divided by the batch size. Unless I'm getting this wrong here you add the number of non-pad tokens in the batch to the denominator. This is available through `label.shape[0]` and also actually what the mxnet [CrossEntropy](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/metric.py#L918) metric does.

How about contributing this back to MXNet in the long term? Ignore pad symbols seems like a generic modification. We'd just need to remove the dependency on LossConfig, which should be easy. We don't need to do this now, but could add a `TODO` to not forget.

Reply:
By default, metrics add up the loss and number of instances (tokens, with or without ignoring padding based on the metric).  The special case here is for _not_ doing this when we've already normalized, instead adding the loss and incrementing the number of examples by 1 since the loss is already divided by the number of tokens.

Good idea contributing this to MXNet--I'll add a TODO.

Reply:
Yeah I get that it makes sense to divide by one in the case `normalize=True`. My point was that in the other case you should divide by the number of elements in the batch, as is done in the corresponding [CrossEntropy](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/metric.py#L918) metric in mxnet, as this is effectively the loss that is optimized when `normalize=False`. So something like:
```
self.num_inst += label.shape[0]
```
This way the training loss and the metric would be consistent.

Reply:
`shape[0]` should be equivalent to `size` after calling `ravel()`.  Are you suggesting that we don't subtract the number of padding tokens here so we're consistent with MXNet?